Stack overview during deps walk with a particular dependency

When I have this dependency in one of my subprojects:

```groovy
compile 'com.jcabi:jcabi-github:0.28'
```

the plugin throws `java.lang.StackOverflowError` (see below). As long as the dependency is removed, everything works just fine. Not exactly sure what's the culprit just yet.

```
Caused by: java.lang.StackOverflowError
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils$_walkDeps_closure6.doCall(KarafUtils.groovy:47)
	at com.github.lburgazzoli.gradle.plugin.karaf.KarafUtils.walkDeps(KarafUtils.groovy:42)
```

--- a/src/main/groovy/com/github/lburgazzoli/gradle/plugin/karaf/KarafUtils.groovy
+++ b/src/main/groovy/com/github/lburgazzoli/gradle/plugin/karaf/KarafUtils.groovy
@@ -34,17 +34,25 @@ class KarafUtils {
 
     static void walkDeps(List<Configuration> configurations, Closure closure) {
         configurations.each {
-            walkDeps(it, it.incoming.resolutionResult.root, closure)
+            Set<ResolvedDependencyResult> alreadyKnownDependencies = new HashSet<>();
+            walkDeps(it, it.incoming.resolutionResult.root, alreadyKnownDependencies, closure)
         }
     }
 
-    static void walkDeps(Configuration configuration, ResolvedComponentResult root, Closure closure) {
+    static void walkDeps(Configuration configuration,
+                         ResolvedComponentResult root,
+                         Set<ResolvedDependencyResult> alreadyKnownDependencies,
+                         Closure closure) {
         root.dependencies.findAll {
             it instanceof ResolvedDependencyResult
         }.collect {
             (ResolvedDependencyResult) it
         }.each {
-            walkDeps(configuration, it.selected, closure)
+            def result = (ResolvedDependencyResult) it
+            if (!alreadyKnownDependencies.contains(result)) {
+                alreadyKnownDependencies.add(result)
+                walkDeps(configuration, result.selected, alreadyKnownDependencies, closure)
+            }
         }
 
         closure(configuration, root)

--- a/src/test/groovy/com/github/lburgazzoli/gradle/plugin/karaf/features/KarafFeaturesTest.groovy
+++ b/src/test/groovy/com/github/lburgazzoli/gradle/plugin/karaf/features/KarafFeaturesTest.groovy
@@ -71,6 +71,31 @@ class KarafFeaturesTest extends KarafTestSupport {
             feature.configurations[0] == project.configurations.runtime
     }
 
+    def 'Recursive Dependency'() {
+        given:
+            def project = setupProject('com.lburgazzoli.github', 'gradle-karaf', '1.2.3') {
+                dependencies {
+                    compile 'com.jcabi:jcabi-github:0.28'
+                }
+            }
+            def task = getKarafFeaturesTasks(project)
+        when:
+            def extension = getKarafExtension(project)
+            extension.features {
+                feature {
+                    name = "feature-1"
+                    description = "my feature n1"
+                }
+            }
+        then:
+            task.generateFeatures(extension.features)
+            extension != null
+            extension.features != null
+            extension.features.featureDescriptors != null
+            extension.features.featureDescriptors.empty == false
+            extension.features.featureDescriptors.size() == 1
+    }
+
     def 'Simple Single Project Dependencies'() {
         given:
             def project = setupProject('com.lburgazzoli.github', 'gradle-karaf', '1.2.3') {

